src/object/cli_obj.c

```diff
@@ -1310,148 +1265,133 @@ obj_iod_valid(unsigned int nr, daos_iod_t *iods, bool update)
+obj_req_valid(void *args, int opc, daos_epoch_t *epoch)
 {
-	daos_obj_fetch_t	*args = dc_task_get_args(task);
-	struct obj_auxi_args	*obj_auxi;
-	struct dc_object	*obj;
-	struct dc_obj_shard	*obj_shard;
-	int			 shard;
-	unsigned int		 map_ver;
-	uint64_t		 dkey_hash;
-	daos_epoch_t		 epoch;
-	int			 rc;
-
-	if (args->dkey == NULL || args->dkey->iov_buf == NULL || args->nr == 0
-	    || !obj_iod_valid(args->nr, args->iods, false))
-		D_GOTO(out_task, rc = -DER_INVAL);
+	daos_obj_fetch_t	*f_args;
+	daos_obj_update_t	*u_args;
+	daos_obj_punch_t	*p_args;
+	int			 rc = 0;
 
-	rc = dc_tx_check(args->th, false, &epoch);
-	if (rc) {
-		if (rc != -DER_INVAL)
-			goto out_task;
-		/* FIXME: until distributed transaction. */
-		epoch = DAOS_EPOCH_MAX; /* = daos_ts2epoch();*/
-		D_DEBUG(DB_IO, "set epoch "DF_U64"\n", epoch);
-	}
-	D_ASSERT(epoch);
+	switch (opc) {
+	case DAOS_OBJ_RPC_FETCH:
+		f_args = args;
+		if (f_args->dkey == NULL || f_args->dkey->iov_buf == NULL ||
+		    f_args->nr == 0 ||
+		    !obj_iod_valid(f_args->nr, f_args->iods, false))
+			D_GOTO(out, rc = -DER_INVAL);
+
+		rc = dc_tx_check(f_args->th, false, epoch);
+		if (rc) {
+			if (rc != -DER_INVAL)
+				D_GOTO(out, rc);
+			/* FIXME: until distributed transaction. */
+			*epoch = srv_io_dispatch ? DAOS_EPOCH_MAX :
+						   daos_ts2epoch();
+			D_DEBUG(DB_IO, "set epoch "DF_U64"\n", *epoch);
+			rc = 0;
+		}
+		break;
+	case DAOS_OBJ_RPC_UPDATE:
+		u_args = args;
+		if (u_args->dkey == NULL || u_args->dkey->iov_buf == NULL ||
+		    u_args->nr == 0 ||
+		    !obj_iod_valid(u_args->nr, u_args->iods, true))
+			D_GOTO(out, rc = -DER_INVAL);
+
+		rc = dc_tx_check(u_args->th, true, epoch);
+		if (rc) {
+			if (rc != -DER_INVAL)
+				D_GOTO(out, rc);
+			/* FIXME: until distributed transaction. */
+			*epoch = srv_io_dispatch ? DAOS_EPOCH_MAX :
+						   daos_ts2epoch();
+			D_DEBUG(DB_IO, "set epoch "DF_U64"\n", *epoch);
+			rc = 0;
+		}
+		break;
+	case DAOS_OBJ_RPC_PUNCH:
+	case DAOS_OBJ_RPC_PUNCH_DKEYS:
+	case DAOS_OBJ_RPC_PUNCH_AKEYS:
+		p_args = args;
+		rc = dc_tx_check(p_args->th, true, epoch);
+		if (rc) {
+			if (rc != -DER_INVAL)
+				D_GOTO(out, rc);
+			/* FIXME: until distributed transaction. */
+			*epoch = srv_io_dispatch ? DAOS_EPOCH_MAX :
+						   daos_ts2epoch();
+			D_DEBUG(DB_IO, "set epoch "DF_U64"\n", *epoch);
+			rc = 0;
+		}
+		break;
+	default:
+		D_ERROR("bad opc %d.\n", opc);
+		D_GOTO(out, rc = -DER_INVAL);
+		break;
+	};
 
-	obj = obj_hdl2ptr(args->oh);
-	if (obj == NULL)
-		D_GOTO(out_task, rc = -DER_NO_HDL);
+out:
+	return rc;
+}
 
-	obj_auxi = tse_task_stack_push(task, sizeof(*obj_auxi));
-	obj_auxi->opc = DAOS_OBJ_RPC_FETCH;
-	rc = tse_task_register_comp_cb(task, obj_comp_cb, &obj,
-				       sizeof(obj));
-	if (rc != 0) {
-		/* NB: process_rc_cb() will release refcount in other cases */
-		obj_decref(obj);
-		D_GOTO(out_task, rc);
-	}
+/* Query the obj request's targets */
+static int
+obj_req_tgts_get(struct dc_object *obj, enum obj_rpc_opc opc,
```

Comment:
based on our code convention, it should be obj_req_get_tgts

Reply:
ok